### PR TITLE
[collective] dynamic shape for send_v2 and recv_v2

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR types
New features

### PR changes
Others

### Describe
send_v2_op (cuda) 与 recv_v2_op (cuda) 修改如下：
修改 attribute，新增dynamic_shape属性（optional，默认值False），作为分布式训练send/recv是否使用动态shape的标识使用。

支持send_v2_op与recv_v2_op的动态shape支持。当设置dynamic_shape属性为True后，send op会首先发送tensor的shape之后再发送tensor。相应的recv op会先接受tensor的shape，并根据接受到的shape修改需要接受的tensor的shape，之后再进行tensor的接受。

support dynamic_shape for send/recv op for gpu training
add a new attr 'dynamic_shape' for both send_v2 and recv_v2, only support gpu with single tensor for now

### CODE REVIEW

File: paddle/fluid/operators/collective/recv_v2_op.cc [hunk 1]
```diff
@@ -44,15 +44,19 @@ class RecvOpV2 : public framework::OperatorWithKernel {
...
-        PADDLE_ENFORCE_GE(out_shape[i], 1,
-                          platform::errors::InvalidArgument(
-                              "The shape attribute for recv_v2 must be set "
-                              "explicitly, but the %dth element is %d which "
-                              "is less than 1.",
-                              i, out_shape[i]));
+      bool dynamic_shape = ctx->Attrs().Get<bool>("dynamic_shape");
+      if (!dynamic_shape) {
+        for (size_t i = 0; i < out_shape.size(); ++i) {
+          PADDLE_ENFORCE_GE(out_shape[i], 1,
```
Comment (gongweibao) [inline paddle/fluid/operators/collective/recv_v2_op.cc:52]: Why must larger than 1?

File: paddle/fluid/operators/collective/recv_v2_op.cu.cc [hunk 2]
```diff
@@ -99,7 +104,66 @@ class RecvOpV2CUDAKernel : public framework::OpKernel<T> {
     auto out_dims = out->dims();
     auto numel = out->numel();
 
-    out->mutable_data<T>(out_dims, place);
+    if (dynamic_shape) {
+      VLOG(3) << "recv_v2 will use dynamic shape with send_v2";
```
Comment (gongweibao) [inline paddle/fluid/operators/collective/recv_v2_op.cu.cc:198]: It's too long.Wrap it as a function?

File: paddle/fluid/operators/collective/recv_v2_op.cu.cc [hunk 3]
```diff
@@ -99,7 +104,66 @@ class RecvOpV2CUDAKernel : public framework::OpKernel<T> {
...
+      ncclDataType_t nccl_dtype =
+          platform::ToNCCLDataType(framework::TransToProtoVarType(shape_dytpe));
+
+      // step1: recv the shape size
+
+      // recv the shape size tensor on gpu
+      framework::Tensor gpu_shape_size_tensor(shape_dytpe);
+      gpu_shape_size_tensor.Resize({1});
+      gpu_shape_size_tensor.mutable_data(place, shape_dytpe);
+      auto *gpu_data = gpu_shape_size_tensor.data<int64_t>();
```
Comment (gongweibao) [inline paddle/fluid/operators/collective/recv_v2_op.cu.cc]: Why int64_t type?

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/collective/recv_v2_op.cc (hunk 1 above)
Comment (FeixLiu) [reply paddle/fluid/operators/collective/recv_v2_op.cc:52]: this is the origin test for send/recv
the send/recv without dynamic shape must have a explicit shape, all dim must be greater than or equal with 1

File: paddle/fluid/operators/collective/recv_v2_op.cu.cc (hunk 2 above)
Comment (FeixLiu) [reply paddle/fluid/operators/collective/recv_v2_op.cu.cc:198]: is there any other place we will reuse this function with all these gpu<->cpu copies?

File: paddle/fluid/operators/collective/recv_v2_op.cu.cc (hunk 3 above)
Comment (FeixLiu) [reply paddle/fluid/operators/collective/recv_v2_op.cu.cc]: no need, int is good 😂

**Review by sandyhouse — DISMISSED** [review]
LGTM

File: paddle/fluid/distributed/collective/ProcessGroupHeter.cc [hunk 4]
```diff
@@ -256,9 +256,9 @@ std::shared_ptr<ProcessGroup::Task> ProcessGroupHeter::Broadcast(
 std::shared_ptr<ProcessGroup::Task> ProcessGroupHeter::Send(
     std::vector<phi::DenseTensor>& in_tensors, int peer) {
 #if defined(PADDLE_WITH_NCCL)
-  PADDLE_ENFORCE_EQ(
-      CheckTensorsInCudaPlace(in_tensors), true,
-      platform::errors::InvalidArgument("All inputs should be in CudaPlace."));
+  if (!CheckTensorsInCudaPlace(in_tensors)) {
+    VLOG(3) << "Inputs not on CUDAPlace for heter send.";
```
Comment (gongweibao) [inline paddle/fluid/distributed/collective/ProcessGroupHeter.cc]: FATAL

File: paddle/fluid/distributed/collective/ProcessGroupHeter.cc [hunk 5]
```diff
@@ -300,9 +301,9 @@ std::shared_ptr<ProcessGroup::Task> ProcessGroupHeter::Send(
 std::shared_ptr<ProcessGroup::Task> ProcessGroupHeter::Recv(
     std::vector<phi::DenseTensor>& out_tensors, int peer) {
 #if defined(PADDLE_WITH_NCCL)
-  PADDLE_ENFORCE_EQ(
-      CheckTensorsInCudaPlace(out_tensors), true,
-      platform::errors::InvalidArgument("All inputs should be in CudaPlace."));
+  if (!CheckTensorsInCudaPlace(out_tensors)) {
+    VLOG(3) << "Inputs not on CUDAPlace for heter recv.";
```
Comment (gongweibao) [inline paddle/fluid/distributed/collective/ProcessGroupHeter.cc]: Same as above.

File: paddle/fluid/operators/collective/recv_v2_op.cu.cc (hunk 2 above)
Comment (gongweibao) [reply paddle/fluid/operators/collective/recv_v2_op.cu.cc:198]: Not only for reusable but also for maintainability.

File: paddle/fluid/operators/collective/recv_v2_op.cu.cc [hunk 6]
```diff
@@ -25,13 +25,98 @@ limitations under the License. */
 namespace paddle {
 namespace operators {
 
+#if (defined(PADDLE_WITH_RCCL) || defined(PADDLE_WITH_NCCL)) && \
+    NCCL_VERSION_CODE >= 2703
+framework::DDim recv_shape_info(const platform::Place &place,
+                                const gpuStream_t &stream,
+                                platform::NCCLComm *comm, const int &peer,
+                                bool with_switch,
```
Comment (wangxicoding) [inline paddle/fluid/operators/collective/recv_v2_op.cu.cc]: 这个参数不需要传，直接通过pg是否nullptr来判断。with_switch这个名字也挺困惑的，感觉用with_group或者has_group好一些

File: paddle/fluid/operators/collective/recv_v2_op.cu.cc [hunk 7]
```diff
@@ -25,13 +25,98 @@ limitations under the License. */
...
+                                distributed::ProcessGroup *pg) {
```
Comment (wangxicoding) [inline paddle/fluid/operators/collective/recv_v2_op.cu.cc]: pg有点不明所以，感觉不如直接group好🤔

File: paddle/fluid/operators/collective/send_v2_op.cu.cc [hunk 8]
```diff
@@ -24,13 +24,90 @@ limitations under the License. */
 namespace paddle {
 namespace operators {
 
+#if (defined(PADDLE_WITH_RCCL) || defined(PADDLE_WITH_NCCL)) && \
+    NCCL_VERSION_CODE >= 2703
+void send_shape_info(const framework::Tensor& x, const platform::Place& place,
+                     const gpuStream_t& stream, platform::NCCLComm* comm,
+                     const int& peer, bool with_switch,
```
Comment (wangxicoding) [inline paddle/fluid/operators/collective/send_v2_op.cu.cc]: 同上

File: paddle/fluid/operators/collective/recv_v2_op.cc [hunk 9]
```diff
@@ -87,6 +93,10 @@ class RecvOpV2Maker : public framework::OpProtoAndCheckerMaker {
         "use_calc_stream",
         "(bool default false) eject CUDA operations to calculation stream.")
         .SetDefault(false);
+    AddAttr<bool>(
+        "dynamic_shape",
```
Comment (wangxicoding) [inline paddle/fluid/operators/collective/recv_v2_op.cc:97]: 另外我觉得可以再给recv_v2 op增加一个可选的ShapeTensor输入，以直接从输入获取shape减少网络传输。

File: paddle/fluid/operators/collective/recv_v2_op.cu.cc (hunk 6 above)
Comment (FeixLiu) [reply paddle/fluid/operators/collective/recv_v2_op.cu.cc]: done

File: paddle/fluid/operators/collective/recv_v2_op.cc (hunk 9 above)
Comment (FeixLiu) [reply paddle/fluid/operators/collective/recv_v2_op.cc:97]: good point, will add later

**Review by wangxicoding — APPROVED** [review]
LGTM

**Review by gongweibao — APPROVED** [review]
LGTM

**Review by Shixiaowei02 — APPROVED** [review]
Paddle inference does not involve changes to the OpMaker in this PR.

**Pull request merged**